datacube/api/query.py

```diff
@@ -181,7 +181,7 @@ def _extract_time_from_ds(ds: Dataset) -> datetime.datetime:
     return normalise_dt(ds.center_time)
 
 
-def query_group_by(group_by='time', **kwargs) -> GroupBy:
+def query_group_by(group_by: str | GroupBy | None = 'time', **kwargs: QueryField) -> GroupBy:
```

Comment:
```suggestion
def query_group_by(group_by: str | GroupBy | None = 'time', **kwargs: QueryDict | None) -> GroupBy:
```
The declared type of `query` in the `load()` function is wrong, it's declared as QueryField but should be `QueryDict | None`.

We could specify `kwargs` as type `Any` here, since it's not used. However, I think we should change the type of this method to only have the `group_by` argument. I think it was written this way for laziness, to not have to extract the `group_by` value in the calling function... not a good reason.

I think this counts as public API, so we should:
- fix the only 2 call sites to not use `**` and just pass the `group_by` value.
- Raise a deprecation warning `if kwargs is not None` and remove it in the next major release.

Reply:
I agree that it seems better to deprecate and fix it properly like you suggest. For now, I've updated the type signatures to align with what you said (and I assume I also found a bug in _spatial.py).

I don't know anything about your API management process, but even if it's technically an API change to update the type annotation, I don't think it's a huge issue to fix an annotation that was previously wrong (client code will still work just as bad or good as before, we're just making it obvious that it's wrong through the type checker).

Reply:
> The declared type of query in the load() function is wrong, it's declared as QueryField but should be QueryDict | None.

`**` args are typed by their elements.   So ` def func(**kwargs: QueryField):`  results in a `kwargs` variable which is of type `dict[str, QueryField]` (i.e. a `QueryDict`).  (Noting that kwargs will never be `None` - if there are no unmatched keyword arguments, kwargs will be an empty dictionary.)

Is there confusion here over this?  Am I missing something? This typing looks consistent and correct to me?

Reply:
Sorry, my mistake! I didn't know how `kwargs` typing worked. A type of `QueryField` is correct in both places.

(I do still think the argument should be removed from `query_group_by` in a separate PR though).